<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

  ---

<center> <font color= #0000> <font size = 5> Ingeniería Financiera | ITESO </font> <br> <br>
   

<center> <font color= #000000> <font size = 4> Microestructura y Sistemas de Trading </font> <br> <br>

<center> <font color=#000000><font size=5> Laboratorio 2 <br> <br> Behavioral Finance en Trading </font> <br> <br>

<center> <font color=#000000><font size=4> I.F. Oscar E. Flores H. | if715029@iteso.mx </font> <br> <br>

  ---

# Abstract

En este trabajo se pretende hacer un análisis de comportamientos en entornos financieros. Se tomarán datos de transaccioones ejecutadas por algún algorítmo o trader (estos mismos proporcionados por Oanda y MetaTrader4), a partir de estos datos generamos un código que analice; los históricos que ha conseguido la cuenta, las medidas de atribuciión al desempeño (que tan bueno es el algorítmo/trader), y finanzas conductuales. 

# Análisis de histórico de cuenta

A continuación se presentará el análisis de una cuenta de trading. Los datos originales son extraidos manualmente de MetaTrader4, estos últimos presentan el historial de transacciones ejecutadas en uno de los servidores de Oanda de la cuenta antes mencionada.

In [13]:
import warnings
warnings.filterwarnings('ignore')
import funciones as fn
import visualizaciones as vz

archivo = "archivo_tradeview_1.csv"
datos = fn.f_leer_archivo(archivo)
datos = fn.f_columnas_tiempos(datos)
datos = fn.f_columnas_pips(datos)
datos = fn.f_capital_acm(datos)
estadisticas = fn.f_estadisticas_ba(datos)
profit_diario = fn.f_profit_diario(datos)
desempenio = fn.f_estadisticas_mad(profit_diario)
sesgos = fn.f_sesgos_cognitivos1(datos)
sesgos2 = fn.f_sesgos_cognitivos2(datos,profit_diario)


Desgargamos datos de traders/algoritmos y hacemos algunos calculos con ellos, entre los cálculos que hemos hecho se encuentra calcular la magnitud de pips según sea el instrumento que se está utilizando, la cantidad de ganancia/pérdida que se ha obtenido después de cerrar la operación, entre otras. A continuación se muestra la tabla de datos generada.

In [14]:
datos.head()

order            opentime type  size  symbol   openprice       s/l  \
0  24316034 2019-08-27 06:57:50  buy  0.04  xauusd  1529.21000  1523.000   
1  24334215 2019-08-27 17:06:20  buy  0.01  eurjpy   117.66400   116.950   
2  24342155 2019-08-27 19:25:39  buy  0.05  eurjpy   117.34600   116.950   
3  24373260 2019-08-28 17:11:33  buy  0.01  eurusd     1.10791     1.104   
4  24392415 2019-08-29 06:35:27  buy  0.01  gbpusd     1.22024     1.208   

         t/p           closetime  closeprice  commission  taxes  swap  profit  \
0  1543.0000 2019-08-27 09:16:01  1533.69000         0.0      0     0   17.92   
1   118.7000 2019-08-29 06:38:16   117.37400         0.0      0     0   -2.73   
2   119.7000 2019-08-28 21:01:24   117.45200         0.0      0     0    5.00   
3     1.1175 2019-08-29 06:38:28     1.10818         0.0      0     0    0.27   
4     1.2370 2019-08-29 06:37:23     1.21990         0.0      0     0   -0.34   

     tiempo  pips  pips_acm  profit_acm  capital_acm         ops  
0    8291.0  44.8      44.8       17.92      5017.92  2019-08-27  
1  135116.0 -29.0      15.8       15.19      5015.19  2019-08-29  
2   92145.0  10.6      26.4       20.19      5020.19  2019-08-28  
3   48415.0   2.7      29.1       20.46      5020.46  2019-08-29  
4     116.0  -3.4      25.7       20.12      5020.12  2019-08-29

### estadísticas básicas de la tabla generada:

In [15]:
estadisticas['df1']

Ops totales              Ganadoras  \
Valor                         84                     69   
Descripcion  Operaciones totales  Operaciones ganadoras   

                                 Ganadoras_c                     Ganadoras_s  \
Valor                                     24                              45   
Descripcion  Operaciones ganadoras de compra  Operaciones ganadoras de venta   

                         Perdedoras                      Perdedoras_c  \
Valor                            15                                15   
Descripcion  Operaciones perdedoras  Operaciones perdedoras de compra   

                                Perdedoras_s  \
Valor                                      0   
Descripcion  Operaciones perdedoras de venta   

                                         Mediana_profit  \
Valor                                             1.205   
Descripcion  Mediana de rendimeintos de las operaciones   

                                   Mediana_pips  \
Valor                                   1252.38   
Descripcion  Mediana de pips de las operaciones   

                                     r_efectividad  \
Valor                                     0.821429   
Descripcion  Ganadoras Totales/Operaciones Totales   

                                      r_proporcion  \
Valor                                          4.6   
Descripcion  Ganadoras Totales/ Perdedoras Totales   

                                              r_efectividad_c  \
Valor                                                0.615385   
Descripcion  Ganadoras Compras/ Operaciones Totales de compra   

                                            r_efectividad_v  
Valor                                                     1  
Descripcion  Ganadoras Ventas/ Operaciones Totales de venta

En este caso podemos observar que son pocas operaciones las que han sido ejecutadas (únicamente 84), aunque realmente la precisión de las operaciones es muy buena, ya que tiene 69 ganadoras (esto equivale al 82.14% de efectividad). Además de esto podemos notar que es prácticamente indiferente entre si ejecutar posiciones de compra o ejecutar posiciones de venta. Se puede notar también que no se tiene  ninguna posición de venta con rendimientos negativos, sin embargo, las posiciones de compra con rendimientos negativos equivalen al 17% de las operaciones totales.

Podemos observar tambien que la mediana de profit es pequeña, aunque positiva. El solo hecho de que la mediana sea positiva nos dice que el algorítmo ha tenido un buen desempeño. Las posiciones en las que ha tenido rendimientos negativos no fueron tan perjudiciales como fueron las posiciones con rendimientos positivos. 

### Rankings:

In [16]:
estadisticas['df2']

btcusd  eurgbp  usdmxn    xauusd    gbpusd  usdjpy    eurusd    audusd  \
rank     1.0     1.0     0.8  0.736842  0.666667     0.6  0.478261  0.333333   

        eurjpy    usdcad  gbpjpy  
rank  0.285714  0.285714    0.25

In [17]:
vz.plot_ranking(estadisticas['df2'])

En la tabla anterior se pretende comparar la tasa de efectividad de cada uno de los instrumentos utilizados. Se puede notar que existen dos instrumentos que tienen 'muy buena tasa' de efectividad, sin embargo, esta tasa se ve sesgada debido a que las muestras son muy reducidas (información que no está contenida explícitamente en la tabla mostrada). Mismo caso con los que aparentemente tienen 'muy mal desempeño' o los instrumentos de extrema derecha.  

# Medidas de atribución al desempeño

Para este apratado se generó una tabla con las medidas de atribución al desempeño más conocidas como son; sharpe ratio, sortino ratios (positivos y negativos), y drawdown, entre otros. 

In [18]:
desempenio

sharpe  sortino_c  sortino_s  \
Valor  0.154555   0.215055   0.131107   

                                             drawdown  \
Valor  de 2019-08-28 a 2019-09-03 drawdown de $190.48   

                                             drawup  information ratio  \
Valor  de 2019-09-03 a 2019-09-25 drawup de $421.83            0.14759   

       avg_rend (annual)  std_rend (annual)  
Valor           0.426481           0.134287

El sharpe es una medida de atribución al desempeño calculada sobre los rendimientos diarios, a estos se les resta la tasa libre de riesgo y se dividen entre la volatilidad del portafolio. Del mismo modo, sortino_c y sortino_v son medidas de desempeño que se calcula en base a las desviaciones estandar de las posiciones en venta y las posiciones en compra de los activos. Utilizando la misma formula de Sharpe se calcula en information ratio, que es el resumen de rendimientos comparados con los presentados por el índice S&P500. Observamos que el Sharpe y el Information Ratio son bastante similares, por lo que podemos asumir que el S&P500 tiene un comportamiento similar a la tasa libre de riesgo. Por último tenemos los rendimientos anualizados y la desviación estandar también anualizada. Para todas las métricas antes señaladas lo deseado es maximizarlas, a excepción de la desviación estándar.

Por último tenemos el drawdown y drawup, que son métricas que señalan que tan mal le va al portafolio cuando le va mal y que tan bien le va cuando le va bien. Para entender un poco mejor este último se presenta la siguiente gráfica: 

In [19]:
vz.plot_profit_diario(profit_diario)

# Behavioral Finance

En este trabajo (propuesto, no fue pedido por el docente) se pretende encontrar sesgos cognitivos. 
Se hace una matriz de probabilidades de Markov. En las filas tenemos la situación "actual" y en las columnas tenemos la probabilidad de seguir en la misma situación o de cambiar a otra situación.

In [20]:
sesgos

lose       win
lose  0.578947  0.421053
win   0.355556  0.644444

Podemos asumir que sí existe un sesgo cognitivo, esto se observa en la probabilidad empírica; perder dinero dado que la última transacción realizada tuvo perdidas tiene una probabilidad del 57.89%, del mismo modo, dada la situación de haber ganado dinero en la última transacción realizada, la probabilidad de ganar dinero en la siguiente transacción será de 64.44%. 

Se puede entonces notar que sí hay un patrón con un poco de sesgo, cuando se gana dinero es muy probable que sigan en rachas de transacciones con rendimientos positivos. Del mismo modo, una vez que entra en 'racha' de perdidas, se mantendrá en ella aunque con menor frecuencia que cuando se tienen rachas de rendimientos positivos. 

In [21]:
sesgos2['ocurrencias']

{'cantidad': 31,
 'detalles': {'ocurrencia_1': {'Timestamp': Timestamp('2019-08-27 09:16:01'),
   'operaciones': {'ganadora': {'instrumento': 'xauusd',
     'volumen': 20,
     'sentido': 'buy',
     'capital_ganadora': 17.92},
    'perdedora': {'instrumento': 'eurjpy',
     'volumen': 20,
     'sentido': 'sell',
     'capital_perdedora': -0.17900000000001626},
    'ratio_cp_capital_acm': -3.56033980360482e-05,
    'ratio_cg_capital_acm': 0.0035643178369046133,
    'ratio_cp_cg': -0.009988839285715192}},
  'ocurrencia_2': {'Timestamp': Timestamp('2019-08-27 09:16:01'),
   'operaciones': {'ganadora': {'instrumento': 'xauusd',
     'volumen': 20,
     'sentido': 'buy',
     'capital_ganadora': 17.92},
    'perdedora': {'instrumento': 'eurusd',
     'volumen': 20,
     'sentido': 'sell',
     'capital_perdedora': -4.0000000000040004e-05},
    'ratio_cp_capital_acm': -7.956066600241467e-09,
    'ratio_cg_capital_acm': 0.0035643178369046133,
    'ratio_cp_cg': -2.2321428571450892e-06}},
  '

In [22]:
vz.plot_sesgos(sesgos2['resultados']['dataframe'])

Como se puede notar en la gráfica de barras, se tienen 4 columnas, de las cuales la primera hace referencia a la cantidad de ocurrencias (31 en total), seguido de esto se observan la proporción de las situaciones en las que se incurre en status_quo (ocasiones en las que la proporción del capital perdedor entre el acumulado, fue menor a la proporción del capital ganador entre el acumulado) 4 de cada 5 veces se llegó al status_quo. Seguido de esto se encuentra la aversión a pérdida, que es la división entre los valores mencionados de Status quo, de tal forma que si la proporción de las pérdidas es 1.5 veces o más la proporción de las ganancias se contabilizará, sino no. Por último se comparan resultados del último activo contra el primero para ver si se tiene (a grandes rasgos) una sensibilidad decreciente (en este caso no se presentó).

Se puede observar que el algorítmo de trading presenta pocos sesgos, no es común que la pérdida flotante supere en 1.5 veces las ganancias de las operaciones ganadoras cerradas, así como también se puede observar que tiene una buena consistencia de ganancias en las operaciones. 

## Conclusiones:

Utilizando algunas de las métricas que se presentan en este documento podemos encontrar comportamientos que pueden ser de utilidad para saber si 'comprar' o 'no comprar' algorítmos de trading como en analizado en este caso. Del mismo modo, se puede utilizar este tipo de análisis para saber si alguno de los trabajadores o ejecutivos del área de trading de alguna empresa en partícular está haciendo bien su trabajo o presenta sesgos conductuales en algunas circunstancias. 